### PR TITLE
Pipeline, Anna Podpryatova - 22FPL2

### PR DESCRIPTION
**Pull request by AnPruch**

### CODE REVIEW

File: lab_6_pipeline/pipeline.py [hunk 1]
```diff
@@ -89,6 +168,10 @@ def analyze(self, texts: list[str]) -> list[StanzaDocument | str]:
         Returns:
             list[StanzaDocument | str]: List of documents
         """
+        new_texts = []
+        for text in texts:
+            new_texts.append(str(self._analyzer(text)))
+        return new_texts
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:174]: rework to comprehension

File: lab_6_pipeline/pipeline.py [hunk 2]
```diff
@@ -78,6 +153,10 @@ def _bootstrap(self) -> AbstractCoNLLUAnalyzer:
         Returns:
             AbstractCoNLLUAnalyzer: Analyzer instance
         """
+        nlp = spacy_udpipe.load_from_path(lang="ru", path=str(UDPIPE_MODEL_PATH))
+        spacy_udpipe.UDPipeModel(lang="ru", path=UDPIPE_MODEL_PATH)
+
+        return AbstractCoNLLUAnalyzer()
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:159]: it is a protocol, it does not have implementation

File: lab_6_pipeline/pipeline.py [hunk 3]
```diff
@@ -78,6 +153,10 @@ def _bootstrap(self) -> AbstractCoNLLUAnalyzer:
         Returns:
             AbstractCoNLLUAnalyzer: Analyzer instance
         """
+        nlp = spacy_udpipe.load_from_path(lang="ru", path=str(UDPIPE_MODEL_PATH))
+        spacy_udpipe.UDPipeModel(lang="ru", path=UDPIPE_MODEL_PATH)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:157]: ?

File: lab_6_pipeline/pipeline.py [hunk 4]
```diff
@@ -239,11 +325,21 @@ def run(self) -> None:
...
+
+
+def main6() -> None:
     """
     Entrypoint for pipeline module.
     """
+    corpus_manager = CorpusManager(path_to_raw_txt_data=ASSETS_PATH)
+    udpipe_analyzer = UDPipeAnalyzer()
+    pipeline = TextProcessingPipeline(corpus_manager, udpipe_analyzer)
+    pipeline.run()
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:341]: you do not need to have separate demonstrations. all marks should be demonstrated in single `main`

File: lab_6_pipeline/pipeline.py [hunk 5]
```diff
@@ -97,6 +180,9 @@ def to_conllu(self, article: Article) -> None:
         Args:
             article (Article): Article containing information to save
         """
+        with open(article.get_file_path(kind=ArtifactType.UDPIPE_CONLLU),
+                  'w', encoding='utf-8') as f:
+            self._analyzer.write(split_by_sentence(article.text), 'conllu')
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:185]: you need to use library-related API. split by sentence is not required. is this method pointed in docs?

File: lab_6_pipeline/pipeline.py [hunk 6]
```diff
@@ -52,11 +117,20 @@ def __init__(
...
+        self._corpus = corpus_manager
+        self.analyzer = analyzer
 
     def run(self) -> None:
         """
         Perform basic preprocessing and write processed text to files.
         """
+        for article in self._corpus.get_articles().values():
+            from_meta(article.get_meta_file_path(), article)
+            from_raw(article.get_raw_text_path(), article)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:129]: aren't they already initialized with `from_raw` in corpus manager?

File: lab_6_pipeline/pipeline.py [hunk 7]
```diff
@@ -52,11 +117,20 @@ def __init__(
...
         """
+        self._corpus = corpus_manager
+        self.analyzer = analyzer
 
     def run(self) -> None:
         """
         Perform basic preprocessing and write processed text to files.
         """
+        for article in self._corpus.get_articles().values():
+            from_meta(article.get_meta_file_path(), article)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:128]: why do we need it?

File: lab_6_pipeline/pipeline.py [hunk 8]
```diff
@@ -17,16 +39,58 @@ def __init__(self, path_to_raw_txt_data: pathlib.Path) -> None:
...
+            if index + 1 != get_article_id_from_filepath(meta) \
+                    or index + 1 != get_article_id_from_filepath(raw)\
+                    or meta.stat().st_size == 0 or raw.stat().st_size == 0:
+                raise InconsistentDatasetError
 
     def _scan_dataset(self) -> None:
         """
         Register each dataset entry.
         """
+        files = [file for file in self.path_to_raw_txt_data.iterdir()
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:83]: use glob with pattern instead of condition

File: lab_6_pipeline/pipeline.py [hunk 9]
```diff
@@ -17,16 +39,58 @@ def __init__(self, path_to_raw_txt_data: pathlib.Path) -> None:
...
+                   if 'raw.txt' in file.name]
+        self._storage = {
+            get_article_id_from_filepath(file):
+            from_raw(
+                path=file,
+                article=Article(url=None,
+                                article_id=get_article_id_from_filepath(file))
+            )
+            for file in files
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:92]: use glob here

File: lab_6_pipeline/pipeline.py [hunk 10]
```diff
@@ -239,11 +325,21 @@ def run(self) -> None:
...
     """
+    corpus_manager = CorpusManager(path_to_raw_txt_data=ASSETS_PATH)
+    udpipe_analyzer = UDPipeAnalyzer()
+    pipeline = TextProcessingPipeline(corpus_manager, udpipe_analyzer)
+    pipeline.run()
 
 
 if __name__ == "__main__":
-    main()
+    main6()
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:345]: .

File: lab_6_pipeline/pipeline.py [hunk 11]
```diff
@@ -17,16 +39,58 @@ def __init__(self, path_to_raw_txt_data: pathlib.Path) -> None:
...
+            raise FileNotFoundError
+
+        if not self.path_to_raw_txt_data.is_dir():
+            raise NotADirectoryError  # built-in
+
+        files = list(self.path_to_raw_txt_data.iterdir())
+        if not files:
+            raise EmptyDirectoryError
+
+        metas = [file for file in files if 'meta.json' in file.name]
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:62]: glob

File: lab_6_pipeline/pipeline.py [hunk 12]
```diff
@@ -17,16 +39,58 @@ def __init__(self, path_to_raw_txt_data: pathlib.Path) -> None:
...
+        raws = [file for file in files if 'raw.txt' in file.name]
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:63]: .

File: lab_6_pipeline/pipeline.py [hunk 13]
```diff
@@ -17,16 +39,58 @@ def __init__(self, path_to_raw_txt_data: pathlib.Path) -> None:
...
+        if len(metas) != len(raws):
+            raise InconsistentDatasetError
+
+        metas.sort(key=lambda x: int(get_article_id_from_filepath(x)))
+        raws.sort(key=lambda x: int(get_article_id_from_filepath(x)))
+
+        for index, (meta, raw) in enumerate(zip(metas, raws)):
+            if not meta.exists() or not raw.exists():
+                raise FileNotFoundError  # built-in
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:72]: how can it be possible?

**Review by demid5111 — CHANGES_REQUESTED** [review]
nice start!

File: lab_6_pipeline/pipeline.py (hunk 1 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:174]: Ok

File: lab_6_pipeline/pipeline.py (hunk 4 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:341]: Ok

File: lab_6_pipeline/pipeline.py (hunk 8 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:83]: Ok

File: lab_6_pipeline/pipeline.py (hunk 9 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:92]: Ok

File: lab_6_pipeline/pipeline.py (hunk 10 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:345]: Removed

File: lab_6_pipeline/pipeline.py (hunk 11 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:62]: Ok

File: lab_6_pipeline/pipeline.py (hunk 12 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:63]: Ok, glob

File: lab_6_pipeline/pipeline.py (hunk 7 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:122]: I thought we should have all information about the article in the Article object.

File: lab_6_pipeline/pipeline.py (hunk 3 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:159]: Deleted

File: lab_6_pipeline/pipeline.py (hunk 2 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:161]: Yes, reworked

File: lab_6_pipeline/pipeline.py (hunk 5 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:184]: I reworked this place. Split_by_sentence is not pointed in docs, but I need it to not pass string to the analyze function

File: lab_6_pipeline/pipeline.py (hunk 7 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py]: I found the place to add meta in documentation, so deleted this line

File: lab_6_pipeline/pipeline.py [hunk 14]
```diff
@@ -207,6 +332,17 @@ def __init__(
...
+        self._analyzer = analyzer
+        self._node_labels = pos
+
+        self.ideal_graph = DiGraph()
+        self.ideal_graph.add_nodes_from(
+            (index, {'label': label})
+            for index, label in enumerate(self._node_labels)
+        )
+        self.ideal_graph.add_edges_from((index, index + 1)
+                                        for index in range(len(self._node_labels) - 1))
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:345]: it is ok for now. think how you would handle a case of finding a tree by pattern:

```
A
|   \
B     C

```

File: lab_6_pipeline/pipeline.py [hunk 15]
```diff
@@ -242,17 +428,60 @@ def _find_pattern(self, doc_graphs: list) -> dict[int, list[TreeNode]]:
...
         """
         Search for a pattern in documents and writes found information to JSON file.
         """
+        for article in self._corpus.get_articles().values():
+            conllu_doc = self._analyzer.from_conllu(article)
+            graphs = self._make_graphs(conllu_doc)
+            pattern_matches = self._find_pattern(graphs)
+            dict_matches = {}
+            for sentence_id, matches in pattern_matches.items():
+                dict_matches[sentence_id] = [treenode_to_dict(match) for match in matches]
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:462]: comprehension

Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:462]: Redid

File: lab_6_pipeline/pipeline.py [hunk 16]
```diff
@@ -189,6 +294,26 @@ def _count_frequencies(self, article: Article) -> dict[str, int]:
...
+        for conllu_sentence in self._analyzer.from_conllu(article).sentences:
+            for word in conllu_sentence.words:
+                word_feature = word.to_dict()["upos"]
+                if word_feature not in pos_freq:
+                    pos_freq[word_feature] = 0
+                pos_freq[word_feature] += 1
+        return pos_freq
+
+
+def treenode_to_dict(node: TreeNode) -> dict:
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:307]: Thank you, used asdict()

File: lab_6_pipeline/pipeline.py [hunk 17]
```diff
@@ -173,11 +264,25 @@ def __init__(self, corpus_manager: CorpusManager, analyzer: LibraryWrapper) -> N
...
         """
+        self._corpus = corpus_manager
+        self._analyzer = analyzer
 
     def run(self) -> None:
         """
         Visualize the frequencies of each part of speech.
         """
+        for article_id, article in self._corpus.get_articles().items():
+            if article.get_file_path(kind=ArtifactType.STANZA_CONLLU)\
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:275]: Documentation does at stage 5.3

File: lab_6_pipeline/pipeline.py (hunk 14 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:345]: I guess, it depends on the way we give this tree to the Pipeline. If we have the edges, then we use them, if nested dict/tuple, then recursively collect all edges

Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:345]: I guess, it depends on the way we give this tree to the Pipeline. If we have the edges, then we use them, if nested dict/tuple, then recursively collect all edges

File: lab_6_pipeline/pipeline.py [hunk 18]
```diff
@@ -62,11 +123,18 @@ def __init__(
...
+        self.analyzer = analyzer
 
     def run(self) -> None:
         """
         Perform basic preprocessing and write processed text to files.
         """
+        for article in self._corpus.get_articles().values():
+            to_cleaned(article)
+            if self.analyzer:
+                article.set_conllu_info(self.analyzer.analyze(split_by_sentence(article.text)))
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:136]: Ok

File: lab_6_pipeline/pipeline.py [hunk 19]
```diff
@@ -107,6 +184,9 @@ def to_conllu(self, article: Article) -> None:
         Args:
             article (Article): Article containing information to save
         """
+        with open(article.get_file_path(kind=ArtifactType.UDPIPE_CONLLU),
+                  'w', encoding='utf-8') as f:
+            f.writelines(article.get_conllu_info())
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:189]: Redid as write

File: lab_6_pipeline/pipeline.py [hunk 20]
```diff
@@ -139,6 +224,7 @@ def analyze(self, texts: list[str]) -> list[StanzaDocument]:
         Returns:
             list[StanzaDocument]: List of documents
         """
+        return self._analyzer.process([Document([], text=' '.join(texts))])
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:227]: Changed

File: lab_6_pipeline/pipeline.py [hunk 21]
```diff
@@ -189,6 +294,26 @@ def _count_frequencies(self, article: Article) -> dict[str, int]:
         Returns:
             dict[str, int]: POS frequencies
         """
+        pos_freq = {}
+        for conllu_sentence in self._analyzer.from_conllu(article).sentences:
+            for word in conllu_sentence.words:
+                word_feature = word.to_dict()["upos"]
+                if word_feature not in pos_freq:
+                    pos_freq[word_feature] = 0
+                pos_freq[word_feature] += 1
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:303]: I'm not sure that's it, but did with get and comprehension

File: lab_6_pipeline/pipeline.py [hunk 22]
```diff
@@ -99,6 +175,7 @@ def analyze(self, texts: list[str]) -> list[StanzaDocument | str]:
         Returns:
             list[StanzaDocument | str]: List of documents
         """
+        return [f"{self._analyzer(text)._.conll_str}\n" for text in texts]
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:178]: We were told to extract conllu information by this mehod at the seminar, and that it lacks last '\n' at the end of the doc, which is required, so I make it this way

File: lab_6_pipeline/pipeline.py [hunk 23]
```diff
@@ -158,11 +254,12 @@ def from_conllu(self, article: Article) -> CoNLLUDocument:
         Returns:
             CoNLLUDocument: Document ready for parsing
         """
+        return CoNLL.conll2doc(input_file=article.get_file_path(kind=ArtifactType.STANZA_CONLLU))
 
 
 class POSFrequencyPipeline:
     """
-    Count frequencies of each POS in articles, update meta info and produce graphic report.
+    Count frequencies of each POS in a rticles, update meta info and produce graphic report.
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:262]: discard

File: lab_6_pipeline/pipeline.py (hunk 14 above)
Comment (demid5111) [reply lab_6_pipeline/pipeline.py:339]: interesting, demonstrate please how it will work

File: lab_6_pipeline/pipeline.py (hunk 6 above)
Comment (demid5111) [reply lab_6_pipeline/pipeline.py]: and what do you think now?

File: lab_6_pipeline/pipeline.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (demid5111) [reply lab_6_pipeline/pipeline.py:184]: Yes! 🎖️

**Review by demid5111 — APPROVED** [review]
appreciate your efforts. address minor comments. in general, i am OK with this implementation. so, congratulations, you're the first in your group to make pipeline up to 10.

File: lab_6_pipeline/pipeline.py (hunk 6 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py]: I think that it's better to fill article texts in CorpusManager, when we create the storage of articles

File: lab_6_pipeline/pipeline.py (hunk 23 above)
Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:262]: Ok

File: lab_6_pipeline/pipeline.py [hunk 24]
```diff
@@ -107,6 +190,9 @@ def to_conllu(self, article: Article) -> None:
         Args:
             article (Article): Article containing information to save
         """
+        with open(article.get_file_path(kind=ArtifactType.UDPIPE_CONLLU),
+                  'w', encoding='utf-8') as f:
+            f.write(article.get_conllu_info())
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:195]: how do you ensure that there is a newline at the end of conllu file content?

Comment (AnPruch) [reply lab_6_pipeline/pipeline.py:195]: Well, this method is called if we previously used our analizer analyze() method in pipeline.run(). Analyze method adds the missing newline manually, how I described before, so we either have the newline or have nothing to convert to conllu at all and don't call to_conllu() method.

**Pull request closed without merging**